### PR TITLE
bugfix: file.directory clean=True when it requires parent dir

### PR DESCRIPTION
**Pull request by hvnsweeting**

### CODE REVIEW

File: salt/states/file.py [hunk 1]
```diff
@@ -312,18 +313,36 @@ def _gen_keep_files(name, require):
     Generate the list of files that need to be kept when a dir based function
     like directory or recurse has a clean.
     '''
+    def _is_subdir(path, directory):
+        path = os.path.realpath(path)
+        directory = os.path.realpath(directory)
+
+        relative = os.path.relpath(path, directory)
+
+        return (not relative.startswith(os.pardir))
```
Comment (isbm) [inline salt/states/file.py]: I find this method a little bit misleading. From its description it means that if I check "sysconfig" is a subdir of "/etc", it should return me True, while it surely does not. So it is as minimum a wrong name of the function, or a bug.

Comment (hvnsweeting) [inline salt/states/file.py]: yes it's a bad name, it's to check a path is under a dir or not, not to check if that is a dir.

````
import os
def is_subdir(path, directory):
    path = os.path.realpath(path)
    directory = os.path.realpath(directory)

    relative = os.path.relpath(path, directory)

    return (not relative.startswith(os.pardir))

assert(is_subdir('/etc/salt', '/etc') == True)
assert(is_subdir('/usr/bin', '/etc') == False)
````

Above ran successfully.

**Comment by thatch45** [comment]
Go Go Jenkins!

**Comment by jfindlay** [comment]
Go Go Jenkins!

**Comment by thatch45** [comment]
@hvnsweeting this is breaking all the file tests, and therefore also breaks all of Salt, We should look at another way to solve this

**Comment by thatch45** [comment]
Since this one is very old and breaks many things I am going to close it out, we should look for another solution to the problem. @hvnsweeting please file an issue that explains what you are trying to fix here

**Pull request closed without merging**